### PR TITLE
Improvement of matroid realization space code

### PR DESCRIPTION
**Pull request by LukasKuehne**
This PR fixes three previous issues:

1) There was a type issue in a corner case which is now fixed (also reported by Chris Eur).
2) Previously the code assumed that the matroid is simple and it crashed otherwise, as reported in the discussion around this PR: https://github.com/oscar-system/Oscar.jl/pull/5539
3) For matroids of rank at least 4, it could happen that the dimension was too large as the distribution of 1's in the initial matrix got stuck. We fixed this by creating a graph and the 1's are then exactly a spanning forest of this graph as described in Theorem 6.4.7 in Oxley's book.

I added tests of the matroids which used to break either of these points and now work fine.

Parts of the code were written jointly with @danteluber. Let me also mention @bschroter so he's up to date.

PS: The test set "realization spaces as schemes" seems to be broken but I'm not sure why. Maybe @HechtiDerLachs knows more on this?

Closes #5539

## Release notes
- Fix a rare type issue in matroid realization space computation
- Allow non-simple matroids in matroid realization space computation
- Fix an issue where the dimension of the realization space might be too large for matroids of rank at least 4

### CODE REVIEW

**Comment by LukasKuehne** [comment]
Just an update on this PR:

An error related to the underlying scheme also mentioned above also appears when I try to `show` a realization space:

```
julia> F = fano_matroid()
Matroid of rank 3 on 7 elements

julia> RSF = realization_space(F);

julia> show(RSF)
Spec of ERROR: method not implemented for realization spaces of type Oscar.MatroidRealizationSpace{ZZRing, ZZRing}
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:44
 [2] underlying_scheme(RS::Oscar.MatroidRealizationSpace{ZZRing, ZZRing})
   @ Oscar ~/Code/OSCAR/Oscar.jl/experimental/MatroidRealizationSpaces/src/realization_space.jl:136
 [3] OO(X::Oscar.MatroidRealizationSpace{ZZRing, ZZRing})
   @ Oscar ~/Code/OSCAR/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Attributes.jl:45
 [4] _show(io::Base.TTY, X::Oscar.MatroidRealizationSpace{ZZRing, ZZRing})
   @ Oscar ~/Code/OSCAR/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Methods.jl:58
 [5] show(io::Base.TTY, X::Oscar.MatroidRealizationSpace{ZZRing, ZZRing})
   @ Oscar ~/Code/OSCAR/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Methods.jl:51
 [6] show(x::Oscar.MatroidRealizationSpace{ZZRing, ZZRing})
   @ Base ./show.jl:484
 [7] top-level scope
   @ REPL[126]:1

julia> AG = affine_geometry(2,3)
Matroid of rank 3 on 9 elements

julia> RSAG = realization_space(AG);

julia> show(RSAG)
scheme(ERROR: MethodError: no method matching MPolyPowersOfElement(::ZZMPolyRing, ::Vector{Any})
The type `MPolyPowersOfElement` exists, but no method is defined for this combination of argument types when trying to construct it.

Closest candidates are:
  MPolyPowersOfElement(::RingType, ::Vector{RingElemType}) where {RingType<:MPolyRing, RingElemType<:MPolyRingElem}
   @ Oscar ~/Code/OSCAR/Oscar.jl/src/Rings/mpoly-localization_types.jl:57
  MPolyPowersOfElement(::RET) where RET<:MPolyRingElem
   @ Oscar ~/Code/OSCAR/Oscar.jl/src/Rings/mpoly-localizations.jl:17

Stacktrace:
 [1] underlying_scheme(RS::Oscar.MatroidRealizationSpace{ZZRing, MPolyQuoLocRing{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem, MPolyPowersOfElement{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem}}})
   @ Oscar ~/Code/OSCAR/Oscar.jl/experimental/MatroidRealizationSpaces/src/MatroidRealizationSpaces.jl:12
 [2] OO
   @ ~/Code/OSCAR/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Attributes.jl:45 [inlined]
 [3] _show(io::Base.TTY, X::Oscar.MatroidRealizationSpace{ZZRing, MPolyQuoLocRing{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem, MPolyPowersOfElement{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem}}})
   @ Oscar ~/Code/OSCAR/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Methods.jl:76
 [4] show(io::Base.TTY, X::Oscar.MatroidRealizationSpace{ZZRing, MPolyQuoLocRing{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem, MPolyPowersOfElement{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem}}})
   @ Oscar ~/Code/OSCAR/Oscar.jl/src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Methods.jl:51
 [5] show(x::Oscar.MatroidRealizationSpace{ZZRing, MPolyQuoLocRing{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem, MPolyPowersOfElement{ZZRing, ZZRingElem, ZZMPolyRing, ZZMPolyRingElem}}})
   @ Base ./show.jl:484
 [6] top-level scope
   @ REPL[129]:1
```
Maybe @HechtiDerLachs knows more on why this is an issue?

File: experimental/MatroidRealizationSpaces/src/realization_space.jl [hunk 1]
```diff
@@ -810,10 +810,11 @@ function n_new_Sgens(
...
 )
   den = denominator(t)
-  if iszero(total_degree(den))
-    # Constant denominator: polynomial ring hom avoids fraction-field JIT overhead.
+  if is_unit(den)
+    # Unit denominator: polynomial ring hom avoids fraction-field JIT overhead.
     # Prime divisors of f(x=num/den) equal those of f(x=num) since den is a unit.
-    m = hom(R, R, map(v -> v == x ? numerator(t) : v, xs))
+    # Note: a constant denominator is not enough, e.g. 2 is not a unit over ZZ.
+    m = hom(R, R, map(v -> v == x ? divexact(numerator(t), den) : v, xs))
```
Comment (joschmitt) [inline experimental/MatroidRealizationSpaces/src/realization_space.jl:831]: I don't claim to understand what is going on in this function, but this change seems to be inconsistent with the comment in line 815? The same might apply below in line 839.

Also, if this previously gave an error for some corner case, would it make sense to add that case to the tests?

Comment (LukasKuehne) [reply experimental/MatroidRealizationSpaces/src/realization_space.jl:831]: Thanks for your comments. Yes you're right. The comment in line 815 is deprecated and incorrect. I just removed it.

I also added two more test cases that now work which used to break before the bug fix.

**Comment by LukasKuehne** [comment]
> Just an update on this PR:
> 
> An error related to the underlying scheme also mentioned above also appears when I try to `show` a realization space:
> 
> ```
[...]
> ```
> 
> Maybe @HechtiDerLachs knows more on why this is an issue?

The last commit fixes this issue.

**Comment by LukasKuehne** [comment]
I know fixed the broken tests in the matroid realization part as far as I can tell.
Anything else I can do in the meantime on the way to merging @joschmitt  @lkastner @antonydellavecchia ?
Thank you very much for your help.
Lukas

**Comment by joschmitt** [comment]
Somebody who has some understanding of the maths / code should review this. Who could do that?

And then there is the release notes business. I edited the pull request description (the very first comment) to contain release notes based on the list of fixed bugs you gave there. Have a look and change it, if I got something wrong.

@aaruni96 Do I need to put labels for every entry of the release notes in the body or can I just add labels to the pull request (every entry should get the same labels)?

**Comment by aaruni96** [comment]
The labels in the PR are inherited by the individual entries. The changelog script still emits a warning, but that's harmless (I did not think about this case when writing the script / warning. This is a happy side effect of the design.)

**Comment by joschmitt** [comment]
> The labels in the PR are inherited by the individual entries. The changelog script still emits a warning, but that's harmless (I did not think about this case when writing the script / warning. This is a happy side effect of the design.)

Great, thank you!

**Comment by LukasKuehne** [comment]
> Somebody who has some understanding of the maths / code should review this. Who could do that?
> 
> And then there is the release notes business. I edited the pull request description (the very first comment) to contain release notes based on the list of fixed bugs you gave there. Have a look and change it, if I got something wrong.
> 
> @aaruni96 Do I need to put labels for every entry of the release notes in the body or can I just add labels to the pull request (every entry should get the same labels)?

Thank you very much. The release notes look good to me.

**Comment by lkastner** [comment]
@danteluber This looks good to me, did you have a chance yet to take a look at this?

**Comment by danteluber** [comment]
> @danteluber This looks good to me, did you have a chance yet to take a look at this?

I don't see any issues, I think its good to go

**Review by lkastner — APPROVED** [review]
